src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,403 @@
+    }
+
+    // LCOV_EXCL_START
+    if (auto const ret = preflight2(ctx); !isTesSuccess(ret))
+        return ret;
+    // LCOV_EXCL_STOP
+
+    // Validation Batch Signers
+    std::unordered_set<AccountID> batchSigners;
+    if (ctx.tx.isFieldPresent(sfBatchSigners))
+    {
+        STArray const& signers = ctx.tx.getFieldArray(sfBatchSigners);
+
+        // Check that the batch signers array is not too large.
+        if (signers.size() > maxBatchTxCount)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]: "
+                                << "signers array exceeds 8 entries.";
+            return temARRAY_TOO_LARGE;
+        }
+
+        // Add batch signers to the set to ensure all signer accounts are
+        // unique. Meanwhile, remove signer accounts from the set of inner
+        // transaction accounts (`requiredSigners`). By the end of the loop,
+        // `requiredSigners` should be empty, indicating that all inner
+        // accounts are matched with signers.
+        for (auto const& signer : signers)
+        {
+            AccountID const signerAccount = signer.getAccountID(sfAccount);
+            if (signerAccount == outerAccount)
+            {
+                JLOG(ctx.j.trace())
+                    << "BatchTrace[" << parentBatchId << "]: "
+                    << "signer cannot be the outer account: " << signerAccount;
+                return temBAD_SIGNER;
+            }
+
+            if (!batchSigners.insert(signerAccount).second)
+            {
+                JLOG(ctx.j.trace())
+                    << "BatchTrace[" << parentBatchId << "]: "
+                    << "duplicate signer found: " << signerAccount;
+                return temREDUNDANT;
+            }
+
+            // Check that the batch signer is in the required signers set.
+            // Remove it if it does, as it can be crossed off the list.
+            if (requiredSigners.erase(signerAccount) == 0)
+            {
+                JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]: "
+                                    << "no account signature for inner txn.";
+                return temBAD_SIGNER;
+            }
+        }
+
+        // Check the batch signers signatures.
+        auto const sigResult = ctx.tx.checkBatchSign(
+            STTx::RequireFullyCanonicalSig::yes, ctx.rules);
+
+        if (!sigResult)
+        {
+            JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]: "
+                                << "invalid batch txn signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    if (!requiredSigners.empty())
+    {
+        JLOG(ctx.j.trace()) << "BatchTrace[" << parentBatchId << "]: "
+                            << "invalid batch signers.";
+        return temBAD_SIGNER;
+    }
+    return tesSUCCESS;
+}
+
+NotTEC
+Batch::checkSign(PreclaimContext const& ctx)
+{
+    if (auto ret = Transactor::checkSign(ctx); !isTesSuccess(ret))
+        return ret;
```

Comment:
Is there a test case for this codepath? Just with a bad outer signature? Didn't see one